### ISSUE TITLE
Bug when exporting CompoundPath to JSON

### ISSUE THREAD

**Issue by lehni** [paperjs/paper.js#1029]
As reported by Daniel Doinov on the mailing list:

> I stumbled upon an interesting bug that accures when exporting complex paths to JSON. 
>
> Here are the examples: 

![before](https://cloud.githubusercontent.com/assets/213275/14179784/4b15179e-f715-11e5-98d3-b616955d681d.png)
Before export 

![after](https://cloud.githubusercontent.com/assets/213275/14179807/56a32c2c-f715-11e5-9771-81e3091bea6b.png)
After export

> The shape you are looking at is a CompoundPath and no manipulation i do on it fixes the problem. 
>
> Exporting it as .svg produses the correct shape but i cant use that feature beacause i need the JSON for the whole project to work. 

SVG:

[pc034 Stay tough_mountains.svg.zip](https://github.com/paperjs/paper.js/files/197880/pc034.Stay.tough_mountains.svg.zip)

**Comment by lehni** [comment]
And [here the sketch](http://sketch.paperjs.org/#S/VVfLchtHDPyVLV3sVFZbA8wDM3b55FNSeVX5mORAU0yZNiWqKMaJlcq/pxtYyspJwu4MHo1GY/nP1d3mdnf16urdp915++Fqvtoeb2h/3pym+835g0xvpl/w97vz7nbZnnab8+7lix9FdJFUJ5Ol9Z4PabpOSzHr2zTzv9LblJbU+qCpo8LK0sscxwwX4h+cT0uVXOlALTWY2ZKbfiHcPt7KIlL6lOLCGIMnJJXBC0ncqoPXR63dvSVjOCklu7cO59dMKpnbNVdPh+kvMkqbrmUp2vM2LX14JpptROYVlljSrxbC1Mfba6mLMV9/zrroJOcc9QGFLWMiCVxppSpDp1Fn1FO1r7C1OS493qbFRuPTlq0ceDZ59rjKGKVLjlJTcU+avfQizX2NrAGdDWNg0W4InHp3vPLITB2ZtojsAF/MJl3mFG57j8YVY0Gpa6adgffB+2t5wuNWHbSkzaMOBYgCZBoe50XFDKZm5CaIIo5sZcxJ8Tgxl4oSLtalx451V7puhsjqTKI5RAWm6gBWunRWXZg4PHt3pQEYgJgtmiupT46pzNfunhmlkUkbHRJMALiI3XJzbP2wLFZlBK/RCiUB1UqJJCRoDtxIc5Rts6NUy+R/3GwqfKujOXajV3KpoQd+WKvB7FobTZRJFmprwetsAKVV8AU4am9spoLYbFcWb2bRpCTVMAe/WyrPzFrG2swmTT1bMcdQpMoc45LZTFCftG7aZetMi7cmMUUlt2h98NKa1Oc28u8Rxwy9QVq12hpmduxiNsW9aKQOqIeDAKif20BYOFMAX9ko9KCis6kWsrjVwt4ib2dtdTxXy4eexCk5xYxZc6v3VRp4Cq31l9bzqgW1BIGHQ0QLwjaGC1XXyN1q6dHuKl6Zxu3RkgMKqnr2hVSmF2DE2Ow/3RXIUgb/h5uNulMQpNpTd5CBw50pheR7VbrOC5523FXOLxJqxfUUhwd4YaogmmKqh03wWFBDJrFIuwo5cZTMGdppdJ9aYDs/qfCTWEEovDWJ1CMpUnA9Wd6Syw4mmoJ4id1qKKS7mfsYsy1UHiJc1kZLhfOu2piiVhUQWjJ0EDliisUpF7NgNbO8kTmlOAWJwGGIDa2RxQ7IIDWeUVVyAIyQ0HjvSPau5gwGU11LMNQVE+Vctk5fd87FPJA2wGpK67ZxQvcSMgBODFTA/rBK83py01gTvRI0YSNSqBYbzypRdG859gHXFUiWvWsGmaE8oLVzwBbN6BeLySYOE9IbZr4JCdOo3q/mCmGxNDpUgM68dZYcTavrhMYOEI0dIakVpGFFOREyOg9XFMCxkcQsUmpcQfBDcHOs4cZhIemQk3DVMqCQvTgJpaK0m1COWyGvhrmEp9Qp6JjHDOYm8RLQQwZIRIaYQhuUQsaTeDfI8eqoNfHB7ZhfJycoTr4gTU6F8qCssl1jwTfQaHLiU0QKemTtuY2Oq2/SVd+lOdswgN2nt9i6vbNTQTle/iC0OAGcIJ74NNcqjkp3cC11DafJ+5J6iVU4ogklj1jy6vQFpj5lo4S6dixol3b1ZSjrGofQ9Hn9gLgo9sXk1wSlNVgQ0l48B3Q2qA5BP3Cz5uADvyjYl8LBxgaP/GNmBj8kKq6g8wUK5sWJml1U0z9fukcxsly9RE5td76h+41TrEgi0lMLvmvlzubkNNeAmkZMUmn9q82PEghXWncLSWlg1RTbIM/r0vZ+W6hfj7pya7Hya3zcoMz/2YqBXv1S3llzgI6DPfyucp7Wjx3eK/gUyYCxL1z2DV6oYGHxuwLLPHnV2DOQ8d4IVQYIZQaMUvLji29e/3bnn87LH/vD4e3xcDzhI/rF+8Nm++kF3vHT+uPD8Q4P49ju7/vj6fz9u59/esm7cYKvlEdOx4+77XnZ3z4d4mUPsr7abM/7z7sfNl92p2Vzc/P2w/5w89LvX1LR5f74sD/vj3fLl+nbN5OmpxcPuwN87G4Q6nz6cxfx97u/lu3u7rw7PcvgeZiLu9fr4cfj8RZHhX7xI+I9fih8uj/u784PV69+/f3f/wA=) to reproduce the issue.

**Comment by lehni** [comment]
This is caused by the same issue as #590. 

@iconexperience, would you have time to discuss the best approach of handling `CompoundPath` sub-path orientation when they are added? The current approach is really dumb and causes more harm than good. I wonder if we should never reorient them? See here:

```js
        // All children except for the bottom one (first one in list) are set
        // to anti-clockwise orientation, so that they appear as holes, but
        // only if their orientation was not already specified before
        // (= _clockwise is defined).
        for (var i = 0, l = !_preserve && items && items.length; i < l; i++) {
            var item = items[i];
            if (item._clockwise === undefined)
                item.setClockwise(item._index === 0);
        }
```

**Comment by iconexperience** [comment]
So let's first explain what the current algorithm in `CompoundPath.insertChildren()` does and what the ideas behind it (probably) was.

`CompoundPath.insertChildren()` automatically set the orientation of the child paths if it has not been explicitly been set before. The orientation of the first path is be set to clockwise, the orienation of all other paths is set to counter-clockwise.

The idea behind this behaviour was to allow creating paths with holes easily. All that needs to be done is to create a path as the container and the paths for the holes and add them to a `CompoundPath`. This works well if first path is the largest and the following paths define the holes in the first path.

<a href="http://sketch.paperjs.org/#S/pVZtb9pIEP4rI38yieOAr0klaHQf+NrTndqr7kOITht7wNssu2h3HeJG/PfOrMHY0ECrWgLb87bzPPMCr5EWS4zG0ecn9HkZJVFuCn6/vgZ8EcuVQvAGnqWrhJLfEEqzBl8irIQvHZg5CJia5cpUuviHRCAsgkVjJWqPBaxL1DCzM03xyK0OelEUpKKwHKjrzYb8eRYWShQF3IHGNbAqnUqbK4xfcw5sx3CfDYcJ0NdDAlYUsnJjGA2Hm8FkFwFr/IhzfzLI6JaCjN53g2QHMT7JRXk6SPbudJClqYiZX8TyniKkrnr0VuQ+PuM6yno08OmDHZnEvBW6MEsuHzPeFEd4aTQXcF9Nqohr6it8EFt0lfKwfUNdgHQg1FrULkgc9Q4fUcgc/95Hje+5eMmuAEnLYtJQ8dBQ03K8NF/llp5uNxDIUqrCoiaYZyImMJdKTY0yxMksWhhVzKLN/hzG1YsdDt3riaQvXirpa5hXOg/cEB3UqeDxxbNJKyfhvySLV0ZqSoD1cH3RPGTN7Y8E0jS9uB7AK7sCXXNjIWa4DHU0odsHGoZFtSTWXKpQL3xJ0svLjg9fofJ8Ujizp+IrJEH8hHv6kjSCtIZLiCVcwWgAFzC6IYIOPXNDHcS+bRb38iE5tJqT1WfqHOqq265uy2145IfNGSabFlQ1OPSHXZjAWvqS6shKL/UiGPyfK5M/raWj9rRmhdbXvTIcNl3o4bcIHzaEB5sTZMs5xH/xmDXpxgPySW8GjR/Rk1p8Rusw/kX0hXQrJeruUGlKX/URHTVp3nnpZMqgiByNI0LWtUmDNPa2wn2CjWVKe0Q7JTz2oqaP/OTStSxoRV3CDS2T4ZEvz9anin4K7mi2iAFtChqv1mo3EFvrbUhvVjyrKWnjjMJevRsOkq7//pgjYNlPA8s6wDJq9TPoBsfwsj48bfQ3tOYteNk5eK3/Mbxeau2+4s1XKTXpktDs40MOtttw8jsdvotOHm1TN+Tu8w17xZqvmBM8as1n/ChqtAx1yjmEUTswPwmJrwIfq8UC7aSxdd6aJ9xZB5F0093E0+D9yWvcIupZBLzSLXZbrj2zifNfKPAdZP2x7FXvB33xVg0DQZ0F5WZRsktnDO1eaqSUGMkoIP0eX3V0/fpv6K/Vo0XxFPazi8b3D5vv">Here is an example</a> that shows how it works:

![image](https://cloud.githubusercontent.com/assets/2122773/15588910/9658d4f0-2390-11e6-9eb7-2c9859f21602.png)

As you can see the orientation of the paths will always be the same after the children have been added. This works very will for cases with a large first path and subsequent smaller paths that define the holes in the large path.

But problem occur if the largest path is not the first path (or if there are several large paths next to each other, or if there are crossing paths, or if there are more than two paths on to of each other, or...)

Here is an example where the large path (the head of the emoji) is not the first child of the `CompoundPath`, but the left eye is. As you can see the orientation of the left eye is clockwise, while all other paths are oriented counter clockwise. Surprisingly the result with `evenodd` fill rule looks correct, but the result with `nonzero`fill rule is broken. <a href="http://sketch.paperjs.org/#S/pVbfb9pIEP5XRn4yieNgX9NKcFEfeKrU053aq/oQotPGHvA2yy7aXYfQKP/7zayNsaGFq84S2N75+c18M/ASabHCaBJ9fkRfVFESFabk9+trwGexWisEb+BJuloo+R2hMhvwFcJa+MqBWYCAmVmtTa3Lv+gIhEWwaKxE7bGETYUa5nauyR+ZbYNclCWJyC076lu3ih80SaTbJ0BqC2mdh6KSqgyxgeTa+CBTwi7R+YSiyaKi8K5W3oHUrT8BG2v0shW0OUXa6O9ozTwi30qBrSkSORXrtZJYsil/noQF3OJHXHi4BY0b4ETTmbSFwvilYJh2AnfZ23EC2bvxfQJWlLJ2E8jHr6Npz8cnuaxOO8nfnHayMjUhP+lhTB7oq+fhHXlIXf3grSh8fMY0y/umGUcf7aJXKMpfDZ6Nd/m3vbBCl2bFTOLONTwRXhrNXNoTi8jhGqqJpsdt79o31GXoldqIrQsnjmjMIUpZ4J97r/Fd27yk60DSlDEJeO73yYUurcw32WLsE5NAMvMsaoJ5xmMS+DQzylBN5tHSqHIe9YrAuAa+Q9BBkb54qaTfwqLWRagNlYOGBjw+e1bpzunwbzqL10ZqSoPlcH3RPOTN7bcE0jS9uB7BC5sCXQtjIWa4DDWb0u13mstlvaKquVShXvqKTi8vezZ8hc5zpBBzIOIrJEH1Cff0OWkO0i1cQizhCrIRXEB2QwU6tCwMMYhtuyzu5H1yqLUgrc/EHGLV276srW145IfXM5VsKKi24NAfspB2iPQVdZOFXtLSYIV/CmWKx410RE9r1mj9dtCGQ9IFDv+s4OOm4EHnRLHlAuI/eMyadOMR2aQ3o8aOypNafELrMP5F9KV0ayW2/aHSlL4aIjoiadF76WXKoKg4GjNC1tdJw2nsbY37BBvNlDaRdkp4HHhNH/jJpRtZ0pK7hBtaJuMjW56tT7yqb2m2qALalDRendZuIFrt1qU3a57YlKRxTm6v3oxHSd9+H+YIWP6fgeU9YDlR/Qy60TG8fAiv+4X6Mbz8HLzO/hjeILVuX/Hmq5Wa9ovQ7OPDGrTbcPp/GL7zThYdqZvi7vMNe8Wab1gQPKLmE34UW7QMdcY5hFE7UD8Jia8SH+rlEu200XXemkfcaYcj6Wa7iafBe89r3CJq+q/AK91in3JdzMbP19DgW8iHYzno3g948bMehgL1FpSbR8kunQl0e6k5pcTojBzS7/FVTzbs/yv9y3uwKB7DfnbR5O7+9V8=">Here is the sketch</a>

![image](https://cloud.githubusercontent.com/assets/2122773/15589060/67ac2ce6-2391-11e6-86b4-c32a5e9eef71.png)

BTW, in the original example with the mountain, the first part is one of the tiny holes at the right of the mountain:

![image](https://cloud.githubusercontent.com/assets/2122773/15589185/09b35f32-2392-11e6-9026-0e6fbad3b8d7.png)

**Comment by lehni** [comment]
@iconexperience I think what we should ask ourselves is also if we should do any automatic sorting at all when creating `CompoundPath` items / adding children to them...

**Comment by iconexperience** [comment]
@lehni I agree, but I am first trying to understand in which cases the reorientation works where ist causes problems.

Here are <a href="http://sketch.paperjs.org/#S/pVZbb9MwFP4rR3lKtyxrAwPUMvHQJyQQiIF4WCvkJaeNmWtHtrOum/rfObaTNmkZAxGpTXNuPt93LuljJNkKo3F0dYs2L6MkylXhns/PAe/ZqhIIVsEdNzUT/AGhVGuwJULFbGlALYDBVK0qVcviM4mAaQSNSnOUFgtYlyhhpmeS4pHbxutZUZCKwrpAXe/G8L0kDTf7BMhswbWxkJdcFP5sIL1U1usE00s0NqHTeF7S8aYW1gCXTTwGa63kslE0OUVSyQfUahZRbCFA13QSBWVVJTgWztV97piGnOtcoIFLuJ5PnHChNMROw0k2nNDtLbyg2+npAB6dAdC19yQjiWtwCNOpl8SPueNHj+E6Gw4TGL0ezhPQrOC1GcOb4XYwacOEEKlWllmMOZzAKCOP4Edf80NTk1a1KePwEJTbFo2h8nUJj11ePcGjp1ijHLfhEs/PVAlF6c4irTZM3IgaZ9F2EMI3NH+zXHC7gUUtc8uVdBWmUoPFe+tZa+Uk/EqyuFJcUtWcHs5Pwo8s3F4kkKbpyXmHzx7po0A6Fb5eEZMmFSiXtjysgbs89e4kf2ZP5S6fBNXB39P7JAjSDZwCsX0Go4Gj/ILKc+iZKyqh891lcc3nyaHVgqyuaHLGMHrV1XVKvO3V6AkmNZOFWokNGAxdH2aMOT11Prelqq1TWk6t7gx+5ELlt2tuaFq1qlDbTa8MBc/x0z5I7Ef6KcKbLvc2fyCbLyD+6Po8pBsPyCe9GAQ/oifVeIfaYPyP6AtuKsE2HlczxpLSF31ER+2ddx4OR1MoiSNC1rVJvTS2usbOWHnL1BIkI9wQ9jxu3C+TrnlBS+kULmgqh0e+boK+1H4TzCJiQKqimEU7q3YgGusmpFXVB1zYlLSxG/mzl8NB0vXfH3MELPtrYFkHWEat/gy6wTG8rA9vt1d/Dy97Dt7O/xheL7XdVnLrtRZi0iUhvJ4OOWhW2+R/OryNTh67pg7k7vP1e0Wrn5gTPGrNO/zANqgd1KnLwY/agfkfIbmrwJt6uUQ9CbbGanWLrbUXcTNtJ54G7x1xudSIkt5wfnFjt+V2Z4Y4332BLyHrj2Wver/pi6dq6AnqLCgzi5I2nTHs9lKQUmIko4D0Qjzr6Pr139J/kxuN7NbvZxONr+fbXw==">three</a> <a href="http://sketch.paperjs.org/#S/pVZLb9pAEP4rI5/s4Bhwmx6gUQ+cKqVq1bTqIUTVxh7wlmXX2l2H0Ij/3tm1DTbk0aqWwGbe3zczax4DydYYTILrFdqsCOIgU7n7PRwCPrB1KRCsgntuKib4b4RCbcAWCCWzhQG1AAYztS5VJfMvJAKmETQqzVFazGFToIS5nkuKR25br2d5TioK6wJ1vRvDj5I03BwKILMF18ZCVnCR+9xAeqms1wmml2hsTNl4VlB6UwlrgMsmHoONVnLZKJqaAqnkb9RqHlBsIUBXlImCsrIUHHPn6j73TEPGdSbQwCXc3E6dcKE0hE7DSTaa0u09jN19MIjg0VkAXY1bUlamCCVuwEFMZl4aPmaOID2Bm3Q0ioG+bmPQLOeVmVCsEZxT1DN62kWRz7lrKzLUgi5pPnRP8Ohp0ignbQ2xxzhTQlHGeWBKzeVyqRHlPGgTNGR9t1xwu4VFJTPLlXR9ooaBxQfrsbdyEn4jWVgqLol7p4fhWf2Q1rc3MSRJcjbskNKjblxTR+2r1kSHSQTKpS2OiXSX589l8jl7Knf5IohMf08e4lqQbGEAIScux5Ej84I4PvbMFPXB+e6ruOG38bHVgqyuaf6pN++6ul3NnX/sdekZJjWTuVqLLRisZ7feFOb0NL/cFqqyTmmpQd7gZyZUttpwQzunVYnabnttyHmGnw9BQr+YzxHezKq3eYFsvoDwkxvWutwwIp/kIqr9iJ5E4z1qg+E/os+5KQXbelzNMkoqX/QRnQx41vnRqdSvplASx4Ssa5N4aWh1hYcCa8vEEiQjmMVe1OTOPZlkw3M6WgZwQSs5OvF1O/TVHROXtEPEgFR5Pg/2Vu1CNNZNSKvKK1zYhLRhSmHP346iuOt/SHMCLP1rYGkHWEqj/gq66BRe2oe3Px2fhpe+Bm/vfwqvV9r+XKKsshJi2iWhfskcc9AcbtP/mfA2Onnsh7om91CvP1e0+oUZwaPRvMcrtkXtoM5cDX7VjsxfhOSuHO+q5RL1tLY1VqsVttZexM2s3XhavA/EZXNQgzu6NXZHbp+zjvPDN/gS0v5a9rr3xFw810NPUOeAMvMgbsuZwP5cqqVUGMkoIL3Vzju6fv939A/jTiNb+fPZBJOb290f">more</a> <a href="http://sketch.paperjs.org/#S/pVbfb9MwEP5XTnlKtyxrM4ZQy8RDn5CGQAzEwzohL7k2Zq4d2c66MvV/52wnbdJuwCBSm/R++b677y59jCRbYjSOru7Q5mWURLkq3O/TU8AHtqwEglVwz03NBP+JUKoV2BKhYrY0oObAYKqWlapl8YlEwDSCRqU5SosFrEqUMNMzSfHIbe31rChIRWFdoK53Y/hekoabXQJkNufaWMhLLgp/NpBeKut1gukFGpvQaTwv6XhTC2uAyyYeg5VWctEompwiqeRP1GoWUWwhQNd0EgVlVSU4Fs7Vfe6ZhpzrXKCBC7i+mTjhXGmInYaTbDih21s4o9vx8QAenQHQtfO8dPmRpcQVOJjp1Ivjx9wVSY/hOhsOE6CvmwQ0K3htxnA23AwmbaxOnNRqJo1gFuOYwwmMBnAEb8h9uG9u0qo2Zdzx3VnskrtaMkL/wuSyw+R8nH9MzvsGi01beENM63Ijdgn2BI+eDRrluI2Z+FZOlVCU9yyqUNezaDMIkRsyfLVccLuGeS1zy5V0PCRCgsUH63vbykn4hWRxpbgkbjk9nB6FhyzczhJI0/TotNP1HjVGgRpU+npJ1TSpQLmw5T5T3OXL707yZ/ZU7vJJUC/8PX1IgiBdwzHsCj06pxbte+aK2uh8t1lc85tk32pOVlc032MYve7qOm3e9NrzTCWp/YVaijUYDLMZNgFzeppPbktVW6e0nAbSGXzPhcrvVtzQTtGKWmbXvTYUPMePuyCxXzzPFbyZRW/zm2LzOcQfHNdDuvGAfNLzQfCj8qQa71EbjF+IvuCmEmztcTXLRlL6oo/ogNl558f+AhFK4oiQdW1SL42trjsTHSw7A9jzuHVPJl3xglbnMZzvjWTwdcPz2a3BCxoeqoBURTGLtlbtQDTWTUirqkuc25S0cUZhT14NB0nXf3fMAbDsr4FlHWAZUf0P6AaH8LI+vO32fxpe9id4W/9DeL3UtgvJrdhaiN4GDi/R/Ro0W23yPwxvo5PHltShuLt8/V7R6gfmBI+oeU8viTVqB3XqcvCjtmf+W0juKvC2XixQT4KtsVrdYWvtRdxM24mnwXtHtVxoREnvYbezNXYptz0zxPnmG3wBWX8se917ghfP9dAXqLOgzCxK2nTGsN1LQUqJkYwC0kvxpKPr939D/6BuNbI7v59NNL6+2fwC">examples</a> that show that the problem occur mainly with `nonzero` fill rule.

But an interesting point is, that for `evenodd`fill rule the reorientation is not required at all in the examples above. For `evenodd`fill rule all examples above are displayed correctly independently from the orientation of the child paths. <a href="http://sketch.paperjs.org/#S/pVZtb9pIEP4rI38yieOAr0klaHQf+NrTndqr7kOITht7wNssu2h3HeJG/PfOrMHY0ECrWgLb87bzPPMCr5EWS4zG0ecn9HkZJVFuCn6/vgZ8EcuVQvAGnqWrhJLfEEqzBl8irIQvHZg5CJia5cpUuviHRCAsgkVjJWqPBaxL1DCzM03xyK0OelEUpKKwHKjrzYb8eRYWShQF3IHGNbAqnUqbK4xfcw5sx3CfDYcJ0NdDAlYUsnJjGA2Hm8FkFwFr/IhzfzLI6JaCjN53g2QHMT7JRXk6SPbudJClqYiZX8TyniKkrnr0VuQ+PuM6yno08OmDHZnEvBW6MEsuHzPeFEd4aTQXcF9Nqohr6it8EFt0lfKwfUNdgHQg1FrULkgc9Q4fUcgc/95Hje+5eMmuAEnLYtJQ8dBQ03K8NF/llp5uNxDIUqrCoiaYZyImMJdKTY0yxMksWhhVzKLN/hzG1YsdDt3riaQvXirpa5hXOg/cEB3UqeDxxbNJKyfhvySLV0ZqSoD1cH3RPGTN7Y8E0jS9uB7AK7sCXXNjIWa4DHU0odsHGoZFtSTWXKpQL3xJ0svLjg9fofJ8Ujizp+IrJEH8hHv6kjSCtIZLiCVcwWgAFzC6IYIOPXNDHcS+bRb38iE5tJqT1WfqHOqq265uy2145IfNGSabFlQ1OPSHXZjAWvqS6shKL/UiGPyfK5M/raWj9rRmhdbXvTIcNl3o4bcIHzaEB5sTZAc90ZBSItPd8fFfPHgNgHhAUYbpzS+iL6RbKVF3h0pT+qqP6KhJ885LJ1MGReRoHBGyrk0apLG3Fe4TbCxT2iPaKeGxFzV95CeXrmVBK+oSbmiZDI98ebY+VfRTcEezhc+oTUHj1VrtBmJrvQ3pzYpnNSVtnFHYq3fDQdL13x9zBCz7aWBZB1hGrX4G3eAYXtaHp43+hta8BS87B6/1P4bXS63dV7z5KqUmXRKafXzIwXYbTn6nw3fRyaNt9obcfb5hEqz5ijnBo9Z8xo+iRstQp5xDGLUD85OQ+CrwsVos0E4aW+etecKddRBJtx+5AfzJa9wi6lkEvNItdluuPbOJ818o8B1k/bHsVe8HffFWDQNBnQXlZlGyS2cM7V5qpJQYySgg/R5fdXT9+m/or9WjRfEU9rOLxvcPm+8=">Here is the first example again</a>, but this time it randomly hard wires the orientation of the child paths, so they will not get reoriented by `insertChildren()`. If you run the script several times you will see that for `evenodd` fill rule the result will always look correctly.

So the result of these simple test is

1. The current automatic reorientation of the child paths causes massive problems if `nonzero` fill rule is used.
2. The reorientation is not necessary if `evenodd` fill rule is used.

As a conclusion I would say that the automatic reorientation of the paths should not happen at all when inserting paths into a `CompoundPath`. I currently cannot think of a situation where this would be helpful.

**Comment by lehni** [comment]
Yes that was my conclusion as well. We could then offer the smart `reorient()` for `nonzero` as a method that you call when you're done adding children. And what if we made `evenodd` the default fill-rule? It would help people who expect compound paths to magically do the right thing...

**Comment by iconexperience** [comment]
If you set the default fill rule to `evenodd`, things may become easier because creating compound paths will be more intuitive. On the other hand, if you display the image somewhere else where `nonzero` is the default fill rule, everything will look messed up and you do not understand why.

I have been through this (before working with Paper.JS). All graphics looked great during development, but when we displayed them in a certain browser (was it Firefox?), all looked totally messed up, because we were using `evenodd`, but the browser only supported `nonzero`. 

This is not an easy decision. I personally would stick with `nonzero`.

Regarding the `reorient()` I think this function is awesome and must be separated from `resolveCrossings()`. I am not fully sure if it will work on all cases. We are using it after all crossings have been resolved and there are no crossing paths, but what will happen if there are crossing paths? This should be examined before making it part of the public API.

**Comment by lehni** [comment]
Yes, `reorient()` only works correctly if there are no crossings. It can still be exposed even with that limitation though, as long as it is documented. If there are crossings, the solution is not really defined I would say, as  a path with crossings will be both clockwise and anti-clockwise, in parts. `reorient()` would then just take into account which one of these it is *more*.

**Comment by lehni** [comment]
I am also weary of switching the fill rule. Many older browsers don't do `evenodd` in Canvas... Maybe it's one for v2.0 : )

**Comment by iconexperience** [comment]
`reorient()` is useful in any case, and maybe we can later improve it to handle crossing paths better or even correctly (if that is possible at all). For for now simply using the current code should be sufficient.

Maybe there could be an option to automatically call `reorient()` after inserting children into a `CompoundPath`, but that option should be off by default.

**Comment by lehni** [comment]
I'm against performing such magic behind the scenes... What we currently do is already confusing, e.g. what if you manually created the path, and expect the sequence to be the same after adding?

**Comment by iconexperience** [comment]
Agreed, a change in the sequence would be a real problem.

While we are at this we should think about the behaviour of ìsClockwise()` and `setClockwise()` in `CompoundPath()`. They don't make much sense, either, because the only look at the first child (which can be totally insignificant, as in the mountain svg). Maybe `isClockwise()` should calculate the orientation just like it is done is `Path`, but for all child paths? Or should they even be removed from `CompoundPath`? Are they required?

**Comment by lehni** [comment]
@iconexperience at the moment we use them only in the boolean operations, where we determine how to deal with the operands in relation to the performed type of operation: https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L93
```js
// Give both paths the same orientation except for subtraction
// and exclusion, where we need them at opposite orientation.
if (_path2 && (operator.subtract || operator.exclude)
        ^ (_path2.isClockwise() ^ _path1.isClockwise()))
    _path2.reverse();
```

I agree that it doesn't make that much sense to have CompundPath#isClockwise()... 

If we keep the `sort` parameter that I just added to `reorient()`, then we could just peek at the first child here, since at this point it is save to assume that it's the defining path. Is this the way to go, then?

**Comment by lehni** [comment]
On further thought... We already have `CompoundPath#getArea()` which caches its results, and  `Path#isClockwise()` basically returns `this.getArea() >= 0` already. All we need to do is move it to `CompoundPath`! I think that's the solution, and then we don't need the sorting at all, I believe. Let me investigate.

**Comment by lehni** [comment]
Closed in #a0417040f8d2b8518a3c44915ca1aa1c99c8d784
